Realign Vsphere implementation to controller based impl

prashanth26 opened the pull request:
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
- I still need to test this out but this is my first version of changes. Was testing the alignment of implementation for gRPC based implementations and hence created this PR for your reviews. 
- Also, this PR is to be merged only once MCM 0.30.0 is released. 
- I wasn't too sure of the integration test changes, might need some help here. 

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy
- target_group:   user|operator|developer
-->
```improvement user
Realign Vsphere implementation from gRPC to controller based OOT approach
```

prashanth26 commented:
/hold don't merge yet.

prashanth26 commented:
I just a simple test against a Gardener vSphere cluster and creation and deletion machines. It seemed to work fine.

prashanth26 commented on go.mod:
```diff
@@ -3,19 +3,30 @@ module github.com/gardener/machine-controller-manager-provider-vsphere
+
+replace (
+	github.com/gardener/machine-controller-manager => github.com/prashanth26/machine-controller-manager v0.0.0-20200512162209-f074b63d505f
+	github.com/onsi/gomega => github.com/onsi/gomega v1.5.0
```
There is no use of an older Gomega version. I have removed it from this list as dependency. Thanks for pointing it out.

prashanth26 commented on go.mod:
```diff
@@ -3,19 +3,30 @@ module github.com/gardener/machine-controller-manager-provider-vsphere
+	github.com/gardener/machine-controller-manager => github.com/prashanth26/machine-controller-manager v0.0.0-20200512162209-f074b63d505f
+	github.com/onsi/gomega => github.com/onsi/gomega v1.5.0
+	github.com/prometheus/client_golang => github.com/prometheus/client_golang v0.9.2
+	k8s.io/api => k8s.io/api v0.0.0-20190918155943-95b840bb6a1f // kubernetes-1.16.0
```
Hi Martin,

Do you mean something like this - https://github.com/gardener/machine-controller-manager-provider-vsphere/pull/1/commits/ed78cf8a3058246ba905372bfca8fde158fcb5cb? And also why this suggestion? Is it because of the go deps handles tags with v prefix better?

MartinWeindel replied:
It's only for consistency with other gardener projects.

prashanth26 replied:
Okay makes sense. Thanks for that. I hope these changes are fine.

prashanth26 commented:
> What are the next steps? I see some TODOs in the main package.

I have removed the TODO on the main package. It was copied over from an older commit on MCM. 

Once the PR on MCM is merged sometime this week and release is created, you may merge this after one last testing. Will update you here. You can keep this open until then

prashanth26 commented:
Hi @MartinWeindel ,

I have fixed the issue with the finalizers which we had discussed offline earlier. I have updated this PR with those changes. PTAL as and when you can to test it with your setup.

And if this works well for you, we will make a release of MCM with OOT support soon. Thanks.

Regards,
Prashanth

prashanth26 commented:
Hi All,

I have re-vendored the MCM libraries to the latest release. Now the OOT is supported with the latest [MCM release v0.30.0](https://github.com/gardener/machine-controller-manager/releases/tag/v0.30.0). This PR helps this implementation align with the same.

Thanks,
Prashanth

MartinWeindel commented:
/lgtm
